Add opt in manual remove device blog (#1227)
home-assistant/developers.home-assistant: blog/2022-02-25-opt-in-remove-device.md

---
author: Martin Hjelmare
authorURL: https://github.com/MartinHjelmare
authorTwitter: MartinHjelmare
title: "Opt in support to manually remove device"
---

> In the Home Assistant Core 2022.3 release, [Erik Montnemery](https://github.com/emontnemery) added [general support](https://github.com/home-assistant/core/pull/66188) for integrations to opt in to show a "remove device" button on the device page in the Home Assistant GUI.

Reviewer: ```suggestion
In the Home Assistant Core 2022.3 release, [Erik Montnemery](https://github.com/emontnemery) added [support](https://github.com/home-assistant/core/pull/66188) for any integration to opt in to show a "remove device" button on the device page in the Home Assistant interface.
```

> In the Home Assistant Core 2022.3 release, [Erik Montnemery](https://github.com/emontnemery) added [support](https://github.com/home-assistant/core/pull/66188) for any integration to opt in to show a "remove device" button on the device page in the Home Assistant interface.

Author: ```suggestion
In the Home Assistant Core 2022.3 release, [Erik Montnemery](https://github.com/emontnemery) added [support](https://github.com/home-assistant/core/pull/66188) for any integration to opt in to show a "delete device" button on the device page in the Home Assistant interface.
```

## Implementation

> Integrations that want to do this can implement a callback coroutine function `async_remove_config_entry_device` in the integration `__init__.py` module.

Reviewer: ```suggestion
To opt-in, integrations need to add the function `async_remove_config_entry_device` to their `__init__.py` module.
```

```py
async def async_remove_config_entry_device(
    hass: HomeAssistant, config_entry: ConfigEntry, device_entry: DeviceEntry
) -> bool:
    """Remove a config entry from a device."""
```

> When the user clicks the remove device button for the device and confirms it, the callback will be awaited and if `True` is returned, the config entry will be removed from the device. If it was the only config entry of the device, the device and all its entities will be removed from the device and entity registries.

Author: ```suggestion
When the user clicks the delete device button for the device and confirms it, the callback will be awaited and if `True` is returned, the config entry will be removed from the device. If it was the only config entry of the device, the device and all its entities will be removed from the device and entity registries.
```
